Replace empty_affine_quantizer with direct dispatch to at::native::empty_affine.. .

kimishpatel opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #36815 Add channel shuffle op fp32 + quantized.
* **#36814 Replace empty_affine_quantizer with direct dispatch to at::native::empty_affine.. .**
* #36813 Add quantized adaptive avgpool.
* #36812 Move to using MemoryFormat::ChannelsLast for quantized avgpool2d.

From the flamegraph it seems 40% the time we are spending going through the dispatch stack. I think in quantized model where compute can take less time, such overheads become noticeable

Differential Revision: [D21093840](https://our.internmc.facebook.com/intern/diff/D21093840/)

kimishpatel committed:
Replace empty_affine_quantizer with new_qtensor_cpu.



Differential Revision: [D21093840](https://our.internmc.facebook.com/intern/diff/D21093840/)

[ghstack-poisoned]

dreiss commented:
Can you update the commit message to explain the motivation for this change?

kimishpatel commented:
> Can you update the commit message to explain the motivation for this change?

Updated.

jerryzh168 commented on aten/src/ATen/native/quantized/cpu/qadd.cpp:
```diff
@@ -138,12 +139,14 @@ Tensor qnnpack_add(Tensor qa, Tensor qb, double scale, int64_t zero_point) {
   const auto b_scale = qb_contig.q_scale();
 
-  Tensor qy = at::_empty_affine_quantized(
+  Tensor qy = at::new_qtensor_cpu(
```
could you use at::native:: empty_affine_quantized?

kimishpatel replied:
Whey do you think that would be better? Does it get around the dispatch overhead?

jerryzh168 replied:
It is better because it has the same API as at::_empty_affine_quantized.
yes, it will get around the dispatch overhead.

kimishpatel replied:
Sure. Let me try.

kimishpatel replied:
This is still slower, but not a whole lot. So I think it can make for a decent compromise.

kimishpatel committed:
Update on "Replace empty_affine_quantizer with new_qtensor_cpu."


From the flamegraph it seems 40% the time we are spending going through the dispatch stack. I think in quantized model where compute can take less time, such overheads become noticeable

Differential Revision: [D21093840](https://our.internmc.facebook.com/intern/diff/D21093840/)

[ghstack-poisoned]

kimishpatel committed:
Update on "Replace empty_affine_quantizer with new_qtensor_cpu."


From the flamegraph it seems 40% the time we are spending going through the dispatch stack. I think in quantized model where compute can take less time, such overheads become noticeable

Differential Revision: [D21093840](https://our.internmc.facebook.com/intern/diff/D21093840/)

[ghstack-poisoned]

kimishpatel committed:
Update on "Replace empty_affine_quantizer with new_qtensor_cpu."


From the flamegraph it seems 40% the time we are spending going through the dispatch stack. I think in quantized model where compute can take less time, such overheads become noticeable

Differential Revision: [D21093840](https://our.internmc.facebook.com/intern/diff/D21093840/)

[ghstack-poisoned]

jerryzh168 commented on aten/src/ATen/native/quantized/cpu/qadd.cpp:
```diff
@@ -7,6 +7,7 @@
 #include <ATen/native/quantized/cpu/quantized_ops.h>
 #include <ATen/native/quantized/cpu/init_qnnpack.h>
 #include <ATen/native/quantized/cpu/qnnpack_utils.h>
+#include <c10/core/TensorOptions.h>
```
can these import be removed?

jerryzh168 commented on aten/src/ATen/native/quantized/cpu/qconv.cpp:
```diff
@@ -7,8 +7,10 @@
 #include <ATen/SmallVector.h>
 #include <torch/library.h>
 #include <ATen/cpp_custom_type_hack.h>
+#include <ATen/quantized/Quantizer.h>
 #include <ATen/native/quantized/cpu/fbgemm_utils.h>
 #include <ATen/native/quantized/cpu/qnnpack_utils.h>
+#include <c10/core/TensorOptions.h>
```
can these includes be removed?

jerryzh168 replied:
I mean the added ones

kimishpatel replied:
Yes, I was gonna do that.

kimishpatel committed:
Update on "Replace empty_affine_quantizer with new_qtensor_cpu."


From the flamegraph it seems 40% the time we are spending going through the dispatch stack. I think in quantized model where compute can take less time, such overheads become noticeable

Differential Revision: [D21093840](https://our.internmc.facebook.com/intern/diff/D21093840/)

[ghstack-poisoned]

kimishpatel committed:
Update on "Replace empty_affine_quantizer with new_qtensor_cpu."


From the flamegraph it seems 40% the time we are spending going through the dispatch stack. I think in quantized model where compute can take less time, such overheads become noticeable

Differential Revision: [D21093840](https://our.internmc.facebook.com/intern/diff/D21093840/)

[ghstack-poisoned]

jerryzh168 reviewed (approved):
Looks good

kimishpatel committed:
Update on "Replace empty_affine_quantizer with direct dispatch to at::native::empty_affine.. ."


From the flamegraph it seems 40% the time we are spending going through the dispatch stack. I think in quantized model where compute can take less time, such overheads become noticeable

Differential Revision: [D21093840](https://our.internmc.facebook.com/intern/diff/D21093840/)

[ghstack-poisoned]

kimishpatel committed:
Update on "Replace empty_affine_quantizer with direct dispatch to at::native::empty_affine.. ."


From the flamegraph it seems 40% the time we are spending going through the dispatch stack. I think in quantized model where compute can take less time, such overheads become noticeable

Differential Revision: [D21093840](https://our.internmc.facebook.com/intern/diff/D21093840/)

[ghstack-poisoned]

kimishpatel committed:
Update on "Replace empty_affine_quantizer with direct dispatch to at::native::empty_affine.. ."


From the flamegraph it seems 40% the time we are spending going through the dispatch stack. I think in quantized model where compute can take less time, such overheads become noticeable

Differential Revision: [D21093840](https://our.internmc.facebook.com/intern/diff/D21093840/)

[ghstack-poisoned]
